### ISSUE TITLE
Fail fast

### ISSUE THREAD

**Issue by paulfantom** [unifi-poller/unifi-poller#356]
It would be nice if the poller would fail fast when cannot collect any data from controllers due to configuration errors (including authentication ones). There is no point in running when no data can be collected and failing fast is a good way to make users pay attention.

**Comment by davidnewhall** [comment]
> There is no point in running

Many docker users would disagree. Sometimes poller starts before the controller. In those instances it receives errors from the first or second polls, then it begins working. Would you also want it to die if the controller reboots and becomes unavailable for a few minutes? 

I've been running this app for over 3 years; there's a reason for these choices. What problem are you having that lead you to request this change?

**Comment by paulfantom** [comment]
I will be speaking only about prometheus side of things as this is what I am using.

> Would you also want it to die if the controller reboots and becomes unavailable for a few minutes?

If the poller cannot gather any metrics I would like to have a mechanism that quickly propagates this information in an easy, noticeable way. This can be achieved in multiple ways:
- exiting poller - easy to notice, but can be problematic on issues you pointed out
- reporting non-200 on prometheus scrape - easy to notice as usually there is a catch-all TargetDown alert
- report `unifipoller_up` metric - middle-ground as it needs specific alert to be created

All those look good to me and I ordered them in my personal preference order. However, any of those would be better than none.

> What problem are you having that lead you to request this change?

User error. Rotated credentials on unifi controller and forgot to do the same for unifi-poller. This lead to having no data for an extended period of time.

> I've been running this app for over 3 years; there's a reason for these choices.

I didn't mean to undermine your design choices. Just to improve the application for other scenarios.

---

Just to note, I aim to deploy multiple unifi-pollers (one per unifi controller) and treat them as the main source of data for an alerting system. In such a scenario, I do prefer to have a 1:1 mapping of exporter scrape availability and controller availability. So yes, if controller is down, I would like poller to either be down or reporting non-200 (effectively down from prometheus perspective).

In an ideal world, I would like to have metrics exposed in OpenMetrics format directly by the unifi controller, but we are not in ideal world :(

**Comment by paulfantom** [comment]
> You could use promtail to stick the poller logs into Loki and alert on an error being emitted.

> You may benefit from enabling the web server and using the available data in the API to monitor unifi-poller from an external system. 

I don't think adding more complexity is the way to solve the issue. Plus I don't want to write any more wrappers to something that in most prometheus exporters is done ootb.

> If poller exits, how does your system know when to turn it back on?

In my case it runs in kubernetes, so the scheduler just restarts it. But other process management systems have this feature too. In docker you can set `--restart on-failure`, in systemd you have a similar way with:
```
[Service]
Restart=on-failure
RestartSec=5s
```
I do believe this part is a solved problem in IT. 

> How do you know when poller is not running vs not being able to detect it's not collecting data?

I know when poller is not running because Prometheus reports metric `up == 0` for it. I only know it is not collecting data by knowing beforehand what data is needed to be collected and that's what the issue is about. In prometheus world, this should be the same thing as from alerting perspective both cases are the same since there is no data being collected. I don't want to have a cause when creating alerts, I only care about symptoms (as per https://prometheus.io/docs/practices/alerting/).

> Can you tweak your monitor to detect the latter as well as the former?

No, this is something inherent to prometheus design and is expected to be fixed on the metrics exposition side.

> I feel like there's something just as easy and possibly more useful, than relying on an application to exit when a backend is down.

Yes, there are few methods (mentioned in https://github.com/unifi-poller/unifi-poller/issues/356#issuecomment-858624670). However, in prometheus world all those are on a side of the application exposing metrics and not on the monitoring server. This is one major design difference between prometheus and systems like nagios, sensu, cloudwatch, influxdb, etc. It is this way because prometheus has a built-in way of detecting if the metrics exposing the application is up or not. Relying on this assumption is considered a best practice when it comes to monitoring applications with prometheus.

**Comment by davidnewhall** [comment]
@SuperQ Please enlighten me as to what I'm missing. :)

**Comment by SuperQ** [comment]
I don't think you want to stop/crash the poller on failure. That's not a standard way for exporters to work.

But we do want a signal that can be used with Prometheus to know that the poller is unable to talk to the Unifi controller.

Since the unifi poller handles multiple controllers, it's probably worth using the `unifi_controller_up <bool>` pattern.

**Comment by davidnewhall** [comment]
That sounds like what I'm looking for. Any chance you can point me to an example of this?

**Comment by kquinsland** [comment]
> Since the unifi poller handles multiple controllers, it's probably worth using the `unifi_controller_up <bool>` pattern.

Another common pattern for scrape/exporters is to return the duration of the scrape in addition to the scrape result.
A simple count by result (200/4XX/5XX) is *incredibly* helpful in lots of cases but might be overkill here; bool should be enough for OP.

The duration is usually for more fine-grained  alerting; E.G.: is a scrape that usually takes .845 seconds now trending north of 3 seconds? Consider that a warning that something is about to blow...). OP could use this, also, to indicate a failure. If the scrape takes more than 5s, consider it a failure weather or not the scrape actually returns anything useful after the 5s point.


The [node-exporter](https://github.com/prometheus/node_exporter) will publish metrics like:

```
node_scrape_collector_success{collector="cpu",endpoint="https",instance="someHostNameHere",job="node-exporter",namespace="monitoring",pod="somePodUUIDHerer",service="node-exporter"}
```

Which will either be a `1` or a `0`.

```
node_scrape_collector_duration_seconds{collector="cpu",endpoint="https",instance="someHostNameHere",job="node-exporter",namespace="monitoring",pod="somePodUUIDHerer",service="node-exporter"}
```

Which returns some floating point number like `.00023`

**Comment by SuperQ** [comment]
@kquinsland That metric is useful because the node_exporter performs the collector scrapes in parallel. It lets you see which collector module is the source of slowdowns.

Prometheus already has built-in metrics like `scrape_duration_seconds` and soon `scrape_timeout_seconds`. Most exporters don't need to worry about this.

**Comment by kquinsland** [comment]
> Most exporters don't need to worry about this.

Right. Instrumenting the unifi collector with this would be a bit overkill.  I only mentioned it because there are a few common ways to keep an eye on how a $thing performs... some of which require more fidelity than a single bit :). ... and somebody else asked for an example of implementation patterns.

An app _should_ immediately crash if it's internal state has reached a **dangerous** place; if a stack canary has been corrupted or an undefined/illegal configuration provided or the app is trying to open a file in a non-white-listed directory...for example. If the unifi-poller config is fine and the only issue is that the downstream entity can't be reached... quitting the poller just creates churn/noise in a place adjacent to the actual problem.

If the job name is known, just monitor the `up{instance="unifi-poller.monitoring:9130",job="unifi"}` metric. If the average value of that metric over $someInterval is not 1.0, fire off an alert. This would be enough for OP to detect that something is wrong, but it does require that `unifi-poller` return a `200 OK` only if a **complete** scrape could be performed.

If the controller returned a `403` code, at a **minimum** i would expect that to be logged. I acknowledge that it's overkill for this, but ideally the poller would publish count of status codes exposed as a metric. Parsing the logs to determine that the poller keeps getting 500/400 class errors on one of the controllers is messy; much cleaner / more helpful to get this information directly from the poller.... but it would then mean that relying on the `up{..}` metric is no longer possible.


TL;DR: Lots of ways to measure if something is healthy. If the argument can be made that we need more than a simple "all good" or "at least something is not right" signal then there will be some code that needs to change. Otherwise, poller should try to stay 'alive' as long as possible and only return a 200/OK when scraped _*if and only if*_ a complete scrape of the controller can be done.
